Code/src/main/java/nl/utwente/viskell/ui/components/Block.java

```diff
@@ -169,33 +145,25 @@ protected void propagateConnectionChanges(boolean finalPhase) {
             input.getConnection().ifPresent(c -> c.handleConnectionChangesUpwards(finalPhase));
         }
         
-        // after type checking all input in the first phase, refresh the local expression of this block
-        if (!finalPhase) {
-            this.updateExpr();
-        }
-
         // propagate changes down from the output anchor to connected inputs
         this.getOutputAnchor().ifPresent(a -> a.getOppositeAnchors().stream().forEach(x -> x.handleConnectionChanges(finalPhase)));
 
-        // If the change is not propagated any further down start recomputation.
-        if (finalPhase && this.isBottomMost()) {
-            // Now that the expressions and types are updated, initiate a visual refresh.
-            Platform.runLater(() -> this.staleVisuals.set(true));
```

Comment:
This is a simplification now that the order of invalidating visual state doesn't matter anymore.